rfc/system/0000-caching-of-offloaded-objects.md

```diff
@@ -0,0 +1,161 @@
+    df = pd.Dataframe(...)
+    ...
+    return df
+
+@task(cached=True, version="1.0")
+def bar(df: pd.Dataframe) -> int:
+    ...
+    
+@workflow
+def wf(a: int, b: str):
+    df = foo(a=a, b=b)
+    # Note that:
+    #   1. the return type of `foo` wraps around a pandas datataframe and adds some metadata to it.
+    #   2. the task `bar` is marked as cached and since the dataframe returned by `foo` overrides its hash
+    #      we will check the cache using the dataframe's hash as opposed to the literal representation.
+    v = bar(df=df) 
+```
+
+It's worth noting that this is a strictly opt-in feature, controlled at the level of Type Transformers. In other words, annotating types for which Type Transformers are not marked as opted in will be a no-op.
+
+In the backend, during the construction of the cache key, prior to calling data catalog, in case the `hash` field is set, we will use it to represent the literal, otherwise we will keep [the current behavior](https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/core/literals.proto#L68-L79). A similar reasoning applies to the local executions case.
+
+### Adding the hash to FlyteIDL
+
+The ``Literal`` will contain a new field of type string:
+
+```protobuf
+// A simple value. This supports any level of nesting (e.g. array of array of array of Blobs) as well as simple primitives.
+message Literal {
+    oneof value {
+        // A simple value.
+        Scalar scalar = 1;
+
+        // A collection of literals to allow nesting.
+        LiteralCollection collection = 2;
+
+        // A map of strings to literals.
+        LiteralMap map = 3;
+    }
+
+    // Hash to be used during cache key calculation.
+    string hash = 4;
+}
+```
+
+### Flytekit 
+
+The crux of the mechanics in flytekit revolves around how to expose enough flexibility to allow for arbitrary hash functions to be used, while at the same time providing enough information to flytekit. We propose the use of a `HashMethod` metadata object used in annotated return types. The idea being that during the process of converting from a python value to a literal we apply that hash method and set it in the literal.
+
+We'll lean on [`typing.Annotated`](https://docs.python.org/3/library/typing.html#typing.Annotated) to check the types and annotations. 
+
+Since we cannot assume any hashing characteristics of the hash functions (since we are not talking about perfect hashing in the general case), the `HashMethod` will expose a simple interface, composed of a callable of type `Callable[[T], str]`. That said, we will provide plenty of examples to showcase the flexibility of this approach.
+
+### Cache key calculation
+
+As mentioned above, during the regular cache key calculation, in both the local and the remote cases, flyte takes the input literal map as part of the key. Going forward we will allow for specific literals in that map to use the overridden hash as the representation of that literal.
```

Comment:
Instead of overriding cache key behavior, should we [leverage tags](https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/datacatalog/datacatalog.proto#L248) to generate and attach multiple tags to the same artifact? the benefit is that, lookup by reference will continue to work even against these new hash-annotated-types...

If we have multiple inputs, we can even generate multiple tags with and without hashes... these are purely strings and will be attached to the artifact in a single API call to...

Reply:
TIL. This is way better. Thank you for the suggestion!

Reply:
I updated the RFC to mention the use of multiple tags.